// ngx-onesignal/dist/ngx-onesignal/bundles/ngx-onesignal.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('ngx-onesignal', ['exports', '@angular/core', '@angular/common', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ngx-onesignal'] = {}, global.ng.core, global.ng.common, global.rxjs, global.rxjs.operators));
}(this, (function (exports, i0, i1, rxjs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/interface.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @record
     */
    function OneSignalStub() { }
    if (false) {
        /** @type {?} */
        OneSignalStub.prototype.init;
        /** @type {?} */
        OneSignalStub.prototype.registerForPushNotifications;
        /** @type {?} */
        OneSignalStub.prototype.getUserId;
        /** @type {?} */
        OneSignalStub.prototype.getRegistrationId;
        /** @type {?} */
        OneSignalStub.prototype.getNotificationPermission;
        /** @type {?} */
        OneSignalStub.prototype.setSubscription;
        /** @type {?} */
        OneSignalStub.prototype.isPushNotificationsEnabled;
        /** @type {?} */
        OneSignalStub.prototype.isOptedOut;
        /** @type {?} */
        OneSignalStub.prototype.isPushNotificationsSupported;
        /** @type {?} */
        OneSignalStub.prototype.subscriptionChange;
        /** @type {?} */
        OneSignalStub.prototype.push;
        /** @type {?} */
        OneSignalStub.prototype.on;
        /** @type {?} */
        OneSignalStub.prototype.context;
        /** @type {?} */
        OneSignalStub.prototype.sdkEnvironment;
    }
    /**
     * initialize OneSignal.
     * @see {\@link https://documentation.onesignal.com/docs/web-push-sdk#section--init-}
     * @abstract
     */
    var OneSignalOptions = /** @class */ (function () {
        function OneSignalOptions() {
        }
        return OneSignalOptions;
    }());
    if (false) {
        /** @type {?} */
        OneSignalOptions.prototype.appId;
        /** @type {?} */
        OneSignalOptions.prototype.autoRegister;
        /** @type {?} */
        OneSignalOptions.prototype.allowLocalhostAsSecureOrigin;
        /* Skipping unhandled member: [prop: string]: any;*/
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/decorators.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * @param {?} initProp
     * @return {?}
     */
    function ExecIf(initProp) {
        return ( /**
         * @param {?} target
         * @param {?} name
         * @param {?} descriptor
         * @return {?}
         */function (target, name, descriptor) {
            /** @type {?} */
            var method = descriptor.value;
            descriptor.value = ( /**
             * @return {?}
             */function () {
                if (this[initProp]) {
                    return method.apply(this, arguments);
                }
                return void 0;
            });
        });
    }

    /**
     * @record
     */
    function OnsSignalSubscribed() { }
    if (false) {
        /** @type {?} */
        OnsSignalSubscribed.prototype.isSubscribed;
        /** @type {?} */
        OnsSignalSubscribed.prototype.userId;
        /** @type {?} */
        OnsSignalSubscribed.prototype.registrationId;
        /** @type {?} */
        OnsSignalSubscribed.prototype.notificationPermission;
        /** @type {?} */
        OnsSignalSubscribed.prototype.optedOut;
        /** @type {?} */
        OnsSignalSubscribed.prototype.serviceWorkerActive;
    }
    /** @type {?} */
    var isSubscribe = ( /**
     * @param {?} __0
     * @return {?}
     */function (_b) {
        var isSubscribed = _b.isSubscribed, userId = _b.userId, registrationId = _b.registrationId, notificationPermission = _b.notificationPermission, optedOut = _b.optedOut, serviceWorkerActive = _b.serviceWorkerActive;
        return isSubscribed &&
            Boolean(userId) &&
            Boolean(registrationId) &&
            notificationPermission === 'granted' &&
            optedOut === false &&
            Boolean(serviceWorkerActive);
    });
    var ɵ0 = isSubscribe;
    /** @type {?} */
    var hasSubsctibed = ( /**
     * @param {?} oneSignal
     * @return {?}
     */function (oneSignal) { return Promise.all([
        oneSignal.isPushNotificationsEnabled(),
        oneSignal.getUserId(),
        oneSignal.getRegistrationId(),
        oneSignal.getNotificationPermission(),
        oneSignal.isOptedOut(),
        oneSignal.context.serviceWorkerManager.getActiveState(),
    ]).then(( /**
     * @param {?} __0
     * @return {?}
     */function (_b) {
        var _c = __read(_b, 6), isSubscribed = _c[0], userId = _c[1], registrationId = _c[2], notificationPermission = _c[3], optedOut = _c[4], serviceWorkerActive = _c[5];
        return ({
            isSubscribed: isSubscribed,
            userId: userId,
            registrationId: registrationId,
            notificationPermission: notificationPermission,
            optedOut: optedOut,
            serviceWorkerActive: serviceWorkerActive,
        });
    })); });
    var ɵ1 = hasSubsctibed;
    // @dynamic
    var OneSignalService = /** @class */ (function () {
        /**
         * @param {?} doc
         * @param {?} options
         */
        function OneSignalService(doc, options) {
            this.doc = doc;
            this.options = options;
            this.scriptinitalize = false;
            this.scriptURL = 'https://cdn.onesignal.com/sdks/OneSignalSDK.js';
            this.isSupportedSubject$ = new rxjs.BehaviorSubject(false);
            this.SubscribeStateSubject$ = new rxjs.BehaviorSubject({
                isSubscribed: false,
                userId: '',
                registrationId: '',
                notificationPermission: '',
                optedOut: true,
                serviceWorkerActive: '',
            });
            this.isOptedOutSubject$ = new rxjs.BehaviorSubject(false);
            this.isPushNotificationsEnabledSubject$ = new rxjs.BehaviorSubject(false);
            this.userIdSubject$ = new rxjs.BehaviorSubject(null);
            /**
             * @see {\@link https://documentation.onesignal.com/docs/web-push-sdk#ispushnotificationssupported}
             */
            this.isSupported$ = this.isSupportedSubject$.asObservable();
            /**
             * @see {\@link https://documentation.onesignal.com/docs/web-push-sdk#ispushnotificationsenabled}
             */
            this.isPushNotificationsEnabled$ = this.isPushNotificationsEnabledSubject$.asObservable();
            /**
             * @see {\@link https://documentation.onesignal.com/docs/troubleshooting-web-push#3-check-if-you-are-subscribed}
             */
            this.subscribeState$ = this.SubscribeStateSubject$.asObservable().pipe(operators.shareReplay(1));
            /**
             * @see {\@link https://documentation.onesignal.com/docs/troubleshooting-web-push#3-check-if-you-are-subscribed}
             */
            this.isSubscribe$ = this.subscribeState$.pipe(operators.map(isSubscribe));
            this.userId$ = this.userIdSubject$.asObservable();
            this.init();
        }
        Object.defineProperty(OneSignalService.prototype, "isSupported", {
            /**
             * @return {?}
             */
            get: function () {
                return this.isSupportedSubject$.value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(OneSignalService.prototype, "isSubscribe", {
            /**
             * @return {?}
             */
            get: function () {
                return isSubscribe(this.SubscribeStateSubject$.value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(OneSignalService.prototype, "isInitialized", {
            /**
             * @return {?}
             */
            get: function () {
                return this.scriptinitalize;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(OneSignalService.prototype, "isOptedOut", {
            /**
             * @return {?}
             */
            get: function () {
                return this.isOptedOutSubject$.value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(OneSignalService.prototype, "userId", {
            /**
             * @return {?}
             */
            get: function () {
                return this.userIdSubject$.value;
            },
            enumerable: false,
            configurable: true
        });
        /**
         * @return {?}
         */
        OneSignalService.prototype.subscribe = function () {
            if (this.isSupported) {
                if (this.isOptedOutSubject$.value) {
                    OneSignal.setSubscription(true);
                }
                else {
                    OneSignal.registerForPushNotifications();
                }
            }
        };
        /**
         * call OneSignal.setSubscription(false)
         * @see {\@link https://documentation.onesignal.com/docs/web-push-sdk#section--setsubscription-}
         * @return {?}
         */
        OneSignalService.prototype.unsubscribe = function () {
            if (this.isSubscribe) {
                OneSignal.setSubscription(false);
            }
        };
        /**
         * https://documentation.onesignal.com/docs/web-push-sdk#section-loading-sdk-asynchronously
         * \@example
         * ngxOneSignal.push(['sendTag', 'key', 'value', function(tagsSent) {
         *   // Callback called when tags have finished sending
         * }]);
         * // or
         * ngxOneSignal.push(['sendTag', 'key', 'value']).then((tagsSent) => {
         *   // Callback called when tags have finished sending
         * });
         * @param {?} items push(["functionName", param1, param2]);
         * @return {?}
         */
        OneSignalService.prototype.push = function (items) {
            if (this.isSupported) {
                return OneSignal.push(items);
            }
            else {
                return Promise.resolve();
            }
        };
        /**
         * @private
         * @return {?}
         */
        OneSignalService.prototype.init = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _a_1;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            _b.trys.push([0, 4, , 5]);
                            if (!!this.scriptinitalize) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.addScript()];
                        case 1:
                            _b.sent();
                            return [4 /*yield*/, this.initOneSignal()];
                        case 2:
                            _b.sent();
                            this.scriptinitalize = true;
                            _b.label = 3;
                        case 3: return [3 /*break*/, 5];
                        case 4:
                            _a_1 = _b.sent();
                            this.scriptinitalize = false;
                            return [3 /*break*/, 5];
                        case 5: return [2 /*return*/];
                    }
                });
            });
        };
        /**
         * @private
         * @return {?}
         */
        OneSignalService.prototype.addScript = function () {
            var _this = this;
            return new Promise(( /**
             * @param {?} resolve
             * @param {?} reject
             * @return {?}
             */function (resolve, reject) {
                /** @type {?} */
                var head = _this.doc.getElementsByTagName('head')[0];
                /** @type {?} */
                var script = _this.doc.createElement('script');
                script.type = 'text/javascript';
                script.onload = resolve;
                script.onerror = reject;
                script.src = _this.scriptURL;
                head.appendChild(script);
            }));
        };
        /**
         * @private
         * @return {?}
         */
        OneSignalService.prototype.initOneSignal = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
                var _this = this;
                return __generator(this, function (_m) {
                    switch (_m.label) {
                        case 0: return [4 /*yield*/, OneSignal.init(Object.assign({}, this.options))];
                        case 1:
                            _m.sent();
                            return [4 /*yield*/, OneSignal.on('subscriptionChange', ( /**
                                 * @return {?}
                                 */function () { return __awaiter(_this, void 0, void 0, function () {
                                    var _b, _c;
                                    return __generator(this, function (_d) {
                                        switch (_d.label) {
                                            case 0:
                                                _c = (_b = this.SubscribeStateSubject$).next;
                                                return [4 /*yield*/, hasSubsctibed(OneSignal)];
                                            case 1:
                                                _c.apply(_b, [_d.sent()]);
                                                return [2 /*return*/];
                                        }
                                    });
                                }); }))];
                        case 2:
                            _m.sent();
                            // https://documentation.onesignal.com/docs/web-push-sdk#section-subscription-change
                            _c = (_b = this.isSupportedSubject$).next;
                            return [4 /*yield*/, OneSignal.isPushNotificationsSupported()];
                        case 3:
                            // https://documentation.onesignal.com/docs/web-push-sdk#section-subscription-change
                            _c.apply(_b, [_m.sent()]);
                            _e = (_d = this.isPushNotificationsEnabledSubject$).next;
                            return [4 /*yield*/, OneSignal.isPushNotificationsEnabled()];
                        case 4:
                            _e.apply(_d, [_m.sent()]);
                            _g = (_f = this.isOptedOutSubject$).next;
                            return [4 /*yield*/, OneSignal.isOptedOut()];
                        case 5:
                            _g.apply(_f, [_m.sent()]);
                            _j = (_h = this.SubscribeStateSubject$).next;
                            return [4 /*yield*/, hasSubsctibed(OneSignal)];
                        case 6:
                            _j.apply(_h, [_m.sent()]);
                            _l = (_k = this.userIdSubject$).next;
                            return [4 /*yield*/, OneSignal.getUserId()];
                        case 7:
                            _l.apply(_k, [_m.sent()]);
                            return [2 /*return*/];
                    }
                });
            });
        };
        return OneSignalService;
    }());
    OneSignalService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    OneSignalService.ctorParameters = function () { return [
        { type: Document, decorators: [{ type: i0.Inject, args: [i1.DOCUMENT,] }] },
        { type: OneSignalOptions }
    ]; };
    /** @nocollapse */ OneSignalService.ɵprov = i0.ɵɵdefineInjectable({ factory: function OneSignalService_Factory() { return new OneSignalService(i0.ɵɵinject(i1.DOCUMENT), i0.ɵɵinject(OneSignalOptions)); }, token: OneSignalService, providedIn: "root" });
    __decorate([
        ExecIf('isInitialized'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], OneSignalService.prototype, "subscribe", null);
    __decorate([
        ExecIf('isInitialized'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], OneSignalService.prototype, "unsubscribe", null);
    __decorate([
        ExecIf('isInitialized'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Array]),
        __metadata("design:returntype", void 0)
    ], OneSignalService.prototype, "push", null);
    if (false) {
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.scriptinitalize;
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.scriptURL;
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.isSupportedSubject$;
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.SubscribeStateSubject$;
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.isOptedOutSubject$;
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.isPushNotificationsEnabledSubject$;
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.userIdSubject$;
        /**
         * @see {\@link https://documentation.onesignal.com/docs/web-push-sdk#ispushnotificationssupported}
         * @type {?}
         */
        OneSignalService.prototype.isSupported$;
        /**
         * @see {\@link https://documentation.onesignal.com/docs/web-push-sdk#ispushnotificationsenabled}
         * @type {?}
         */
        OneSignalService.prototype.isPushNotificationsEnabled$;
        /**
         * @see {\@link https://documentation.onesignal.com/docs/troubleshooting-web-push#3-check-if-you-are-subscribed}
         * @type {?}
         */
        OneSignalService.prototype.subscribeState$;
        /**
         * @see {\@link https://documentation.onesignal.com/docs/troubleshooting-web-push#3-check-if-you-are-subscribed}
         * @type {?}
         */
        OneSignalService.prototype.isSubscribe$;
        /** @type {?} */
        OneSignalService.prototype.userId$;
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.doc;
        /**
         * @type {?}
         * @private
         */
        OneSignalService.prototype.options;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/one-signal.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NgxOneSignalModule = /** @class */ (function () {
        function NgxOneSignalModule() {
        }
        /**
         * @param {?=} options
         * @return {?}
         */
        NgxOneSignalModule.forRoot = function (options) {
            return {
                ngModule: NgxOneSignalModule,
                providers: [{ provide: OneSignalOptions, useValue: options }],
            };
        };
        return NgxOneSignalModule;
    }());
    NgxOneSignalModule.decorators = [
        { type: i0.NgModule }
    ];

    /**
     * @fileoverview added by tsickle
     * Generated from: public_api.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * Generated from: ngx-onesignal.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.NgxOneSignalModule = NgxOneSignalModule;
    exports.OneSignalOptions = OneSignalOptions;
    exports.OneSignalService = OneSignalService;
    exports.ɵa = ExecIf;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
